src/core/exec-invoke.c

```diff
@@ -2251,6 +2251,44 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
         return 1;
 }
 
+static int setup_private_pids(const ExecContext *c, ExecParameters *p) {
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        int r, q;
+
+        if (unshare(CLONE_NEWPID) < 0)
+                return log_exec_debug_errno(c, p, errno, "Failed to unshare pid namespace: %m");
+
+        /* The first process created after unsharing a pid namespace becomes PID 1 in the pid namespace, so
+         * we have to fork after unsharing the pid namespace to become PID 1. The parent sends the child
+         * pidref to the manager and exits while the child process continues with the rest of exec_invoke()
+         * and finally executes the actual payload. */
+
+        r = pidref_safe_fork("(sd-pidfd-child)", /*flags=*/ 0, &pidref);
+        if (r < 0)
+                return log_exec_debug_errno(c, p, r, "Failed to fork child into new pid namespace: %m");
+
+        if (r > 0) {
+                /* In the parent process, we send the child pidref to the manager and exit. */
+
+                if (p->pidref_transport_fd >= 0) {
+                        q = send_one_fd_iov(
+                                        p->pidref_transport_fd,
+                                        pidref.fd,
```

Comment:
note that at this point pidfd support is not guaranteed to be available in our tree, this should handle gracefully the case when we have no pidfd to send. (and in that case just send the pid_t, and accept that things are racy)

Reply:
hmmm, so something is wrong here. the receving side in pid1 doesn#t actually expect any fds attached to the message it instead looks at SO_PASSPIDFD instead afaics. but that means it must be confused by which process and is parent and which is the child?

anyway, this seems qutie broken, the sender and the receiver seem to implement different protocols.

Reply:
note that SO_PASSPIDFD is a very recent kernel feature. I think it would be OK to rely on its existance for this systemd feature, given this is a new feature too. But we need to make sure that we fail cleanly on old systems here. And that means that if `pidref.fd` is unset we return an error early, instead of sending an invalid fd.

Reply:
> note that at this point pidfd support is not guaranteed to be available in our tree, this should handle gracefully the case when we have no pidfd to send. (and in that case just send the pid_t, and accept that things are racy)

I think per your point below, we should probably just require PIDFD support to guarantee it won't be racy. In Twine agent, we hit a bunch of racy issues monitoring containers due to PID reuse and its painful.

> hmmm, so something is wrong here. the receving side in pid1 doesn#t actually expect any fds attached to the message it instead looks at SO_PASSPIDFD instead afaics. 

Okay I looked through the manager code and it looks like we use SO_PASSPIDFD to get the PID via ucreds but we also expect a PIDFD to be sent in the message body.

My guess is Daan did this to allow for dealing with kernels without SO_PASSPIDFD. The confusing part is we compare to make sure the ucreds and PIDFD sent via message body match...so it implies we still need SO_PASSPIDFD?

Anyways, yea let me dig more into this, thanks for calling this out.

```
        CMSG_FOREACH(cmsg, &msghdr) {
                if (cmsg->cmsg_level != SOL_SOCKET)
                        continue;

                if (cmsg->cmsg_type == SCM_CREDENTIALS && cmsg->cmsg_len == CMSG_LEN(sizeof(struct ucred))) {
                        assert(!ucred);
                        ucred = CMSG_TYPED_DATA(cmsg, struct ucred);
                } else if (cmsg->cmsg_type == SCM_PIDFD) {
                        assert(pidfd < 0);
                        pidfd = *CMSG_TYPED_DATA(cmsg, int);
                }
        }
```

> note that SO_PASSPIDFD is a very recent kernel feature. I think it would be OK to rely on its existance for this systemd feature, given this is a new feature too. But we need to make sure that we fail cleanly on old systems here. And that means that if pidref.fd is unset we return an error early, instead of sending an invalid fd.

I kinda like this better as opposed to relying on sending a PIDFD twice if SO_PASSPIDFD is enabled. If I go this approach, I can add comments in systemd.exec

Reply:
Okay ended up fixing this - you were right, we were missing parsing SCM_RIGHTS message - and added a comment in manager.c about the protocol